src/type_str.rs

```diff
@@ -0,0 +1,639 @@
+    /// zero-padded on the right. (thus they cannot contain trailing copies of U+0000 'NULL';
+    /// they can, however, contain interior copies)
+    ///
+    /// Like Rust's `char`, the code points must have a value in `[0, 0x110000)`.  However, unlike
+    /// `char`, surrogate code points are allowed.
+    UnicodeStr,
+    /// Code `V`.  Represents a binary blob of `size` bytes.
+    ///
+    /// Can use `Vec<u8>` for serialization.
+    RawData,
+}
+
+impl TypeKind {
+    fn from_char(s: char) -> Option<Self> {
+        match s {
+            'b' => Some(TypeKind::Bool),
+            'i' => Some(TypeKind::Int),
+            'u' => Some(TypeKind::Uint),
+            'f' => Some(TypeKind::Float),
+            'c' => Some(TypeKind::Complex),
+            'm' => Some(TypeKind::TimeDelta),
+            'M' => Some(TypeKind::DateTime),
+            'S' => Some(TypeKind::ByteStr),
+            'U' => Some(TypeKind::UnicodeStr),
+            'V' => Some(TypeKind::RawData),
+            _ => None,
+        }
+    }
+
+    fn to_str(self) -> &'static str {
+        match self {
+            TypeKind::Bool => "b",
+            TypeKind::Int => "i",
+            TypeKind::Uint => "u",
+            TypeKind::Float => "f",
+            TypeKind::Complex => "c",
+            TypeKind::TimeDelta => "m",
+            TypeKind::DateTime => "M",
+            TypeKind::ByteStr => "S",
+            TypeKind::UnicodeStr => "U",
+            TypeKind::RawData => "V",
+        }
+    }
+}
+
+impl TypeKind {
+    // `None` means all sizes are valid.
+    fn valid_sizes(self) -> Option<&'static [u64]> {
+        match self {
+            TypeKind::Bool => Some(&[1]),
+
+            // numpy doesn't actually support 128-bit ints
+            TypeKind::Int |
+            TypeKind::Uint => Some(&[1, 2, 4, 8]),
+
+            // yes, 128-bit floats are supported by numpy
+            TypeKind::Float => Some(&[2, 4, 8, 16]),
+
+            // 4-byte complex numbers are mysteriously missing from numpy
+            TypeKind::Complex => Some(&[8, 16, 32]),
+
+            TypeKind::TimeDelta |
+            TypeKind::DateTime => Some(&[8]),
+
+            // (Note: numpy does support types `|S0` and `|U0`, though for some reason `numpy.save`
+            //        changes them to `|S1` and `|U1`.)
+            TypeKind::ByteStr |
+            TypeKind::UnicodeStr |
+            TypeKind::RawData => None,
+        }
+    }
+
+    /// Returns `true` if `|` endianness is illegal.
+    fn requires_endianness(self, size: u64) -> bool {
+        match self {
+            TypeKind::Bool |
+            TypeKind::Int |
+            TypeKind::Uint |
+            TypeKind::Float |
+            TypeKind::TimeDelta |
+            TypeKind::DateTime |
+            TypeKind::Complex => size != 1,
+
+            TypeKind::UnicodeStr => true,
+
+            TypeKind::ByteStr |
+            TypeKind::RawData => false,
+        }
+    }
+
+    /// Returns `true` if `|` endianness is illegal.
```

Comment:
This comment is copy-pasted from the previous function.

```suggestion
    /// Returns `true` if unit specification is required.
```